src/wfs/src/main/java/org/geoserver/wfs/xml/FeatureTypeSchemaBuilder.java

```diff
@@ -365,20 +369,63 @@ public final XSDSchema buildSchemaInternal(
         return schema;
     }
 
-    private void addAllNamespacesFromCatalog(XSDSchema schema) {
-        WorkspaceInfo localWorkspace = LocalWorkspace.get();
+    private void addRequiredNamespaces(FeatureTypeInfo[] featureTypeInfos, XSDSchema schema) {
+        final WorkspaceInfo localWorkspace = LocalWorkspace.get();
         if (localWorkspace != null) {
             // deactivate workspace filtering
             LocalWorkspace.remove();
         }
-        // add secondary namespaces from the full catalog
         try {
-            for (NamespaceInfo nameSpaceinfo : catalog.getNamespaces()) {
-                if (!schema.getQNamePrefixToNamespaceMap().containsKey(nameSpaceinfo.getPrefix())) {
-                    schema.getQNamePrefixToNamespaceMap()
-                            .put(nameSpaceinfo.getPrefix(), nameSpaceinfo.getURI());
+            if (localWorkspace != null && localWorkspace.isIsolated()) {
+                // Isolated namespace, add only required namespaces from mappings
+                final Map<String, String> schemaNamespacesMap =
+                        schema.getQNamePrefixToNamespaceMap();
+                final List<NamespaceInfo> catalogNamespaces = catalog.getNamespaces();
+                for (FeatureTypeInfo featureTypeInfo : featureTypeInfos) {
+                    Object mapObject =
+                            featureTypeInfo
+                                    .getFeatureType()
+                                    .getUserData()
+                                    .get(Types.DECLARED_NAMESPACES_MAP);
+                    if (!(mapObject instanceof Map)) continue;
```

Comment:
Just to make sure... it a simple feature is in an isolated workspace, and it won't have the map cause it's a simple one, encoding is still going to produce a valid document because simple types have no extra namespace dependencies?

Thinking out loud, wouldn't it make sense to use only the known dependencies regardless of whether the workspace is isolated or not?

Reply:
This method is only used when it is a complex feature case, so simple features namespaces handling is not touched.  The only-on-insolated-workspace condition enforce is for backward compatibility.